### ISSUE TITLE
Resources to support SElinux

### ISSUE THREAD

**Issue by aaronlippold** [inspec/inspec#534]
We should add an selinux resource that can check the status of selinux configs, mode, status and perhaps ones for selinux_rule as well?

**Comment by chris-rock** [comment]
This resource implementation should also be compatible with http://serverspec.org/resource_types.html#selinux and http://serverspec.org/resource_types.html#selinux_module

**Comment by AnotherNerdHere** [comment]
I would like to see this as well.
thanks

**Comment by Paladin** [comment]
Maybe even a resource for the boolean with a syntax like:

describe selinux_boolean 'httpd_can_network_connect' do
  it { should be_true }
end

(or be_on, to make it align with the results of getsebool, if seems better.)

**Comment by erenfro** [comment]
I definitely want this as well. Checking for things like selinux enabled, modules loaded, and even booleans set a specific way would all be quite beneficial. Especially in integration and compliance testing and reporting.

**Comment by erinn** [comment]
+1

**Comment by zbikmarc** [comment]
Any updates on that feature?

**Comment by hassanaftab90** [comment]
Whenever I find selinux disabled in a system, a little part of me dies inside. This should be included in inspec tests.

**Comment by chris-rock** [comment]
Lets define what we want first. Is the following covering the needs?

## InSpec selinux resource

```
describe selinux do
  its('mode') { should eq 'disabled' }
  its('mode') { should eq 'enforcing' }
  its('mode') { should eq 'permissive' }
end

# read all policies
describe selinux.policies do
  its("names") { should eq ['mls'] }
  its("modes") { should_not include 'disabled' }
end

# check multiple policies based on regex
describe selinux.policies(/mls/) do
  its("names") { should eq ['mls'] }
  its("modes") { should_not include 'disabled' }
end

# check that we have not disabled a policy
describe selinux.policies.where { mode == 'disabled'} do
  its("names") { should eq [] }
end

# We use filter table to see all modules
describe selinux.modules.where { status == 'installed' } do
  it { should_not be_enabled }
  it { should_not be_installed }
end

# filter on a single module name
describe selinux.modules('zebra') do
  its("versions") { should cmp '1.13.0' }
  it { should be_enabled }
  it { should be_installed }
end
```

## Serverspec compatibility mode

```
describe selinux do
  it { should be_disabled }
  it { should be_enforcing }
  it { should be_permissive }
end

describe selinux_module('zebra') do
  it { should be_enabled }
  it { should be_installed }
end
```

We are NOT going to support the following Serverspec syntax:
```
describe selinux do
  it { should be_enforcing.with_policy('mls') }
  it { should be_permissive.with_policy('targeted') }
end

describe selinux_module('zebra') do
  it { should be_installed.with_version('1.13.0') }
end
```

**Comment by Paladin** [comment]
Something for the booleans as well. Maybe like:

```
describe selinux_boolean 'httpd_can_network_connect' do
  it { should be_true }
end
```

Pick a syntax (selinux_boolean or selinux.boolean) that's consistent with the rest. If convenient, it'd be great to treat the booleans the same way modules are treated, within limitations (e.g., "selinux.booleans.where {mode == 'on' }" to collect the all booleans that are true) but that's more a "gee it'd be nice" feature rather than a necessary one. As for the actual name of the boolean state, either true/false or on/off or even enabled/disabled would be fine by me (SELinux itself uses on/off, I used "true" in the example simply because be_true is already in inspec).

If the boolean isn't part of the SELinux under test, both condition tests should fail (in the example above, if the SELinux being tested doesn't support the "http_can_network_connect" boolean, for some strange reason, both "be_true" and "be_false" tests should fail -- at the minimum the test should error out because the fact the SELinux we're testing doesn't support something we're testing for needs to be flagged. (That seems obvious, but maybe that's only me.)

**Comment by chris-rock** [comment]
@Paladin Great addition. I think we should stick to the selinux wording to stay consistent. Based on your idea, that would look like:

```
describe selinux_boolean 'httpd_can_network_connect' do
  it { should be_on }
  it { should be_off }
end

# with InSpec filter table
describe selinux.booleans do
  its("names") { should eq ["httpd_can_network_connect", "xen_use_nfs"] }
  its("states") { should eq ["on", "off"]}
end
```
Does this feel right to you?

**Comment by Paladin** [comment]
Going with "state" (as you did) rather than "mode" feels better for a boolean. I used mode in the filter example simply because it was there, but "state" is the better word for it.

So yes, looks good to me.

**Comment by Paladin** [comment]
Just thought of one other possible use-case.

the os resource has things like os.redhat? that can be used to skip tests that don't apply. I'm wondering if an equivalent, such as selinux.installed? could exist to skip tests on systems selinux isn't installed on.

Example: currently I don't install SELinux when building an ubuntu box (might revisit that decision later, but last time I tried it was a nightmare to set up -- don't think I ever got it right). So my SELinux tests look something like:

```
control 'SELnux should be enabled' do
  impact 1.0
  title 'so ensure SELinux is working.'
  desc 'It should make sure SELinux mode is "enforcing".'
  only_if { os.redhat? }
  describe command 'getenforce' do
    it { should exist }
    its('stdout') { should match /Enforcing/ }
  end
end
```

I'm building either Ubuntu or Centos ATM. If I were to throw another distro into my mix, I'd need to come back to every SELinux test batch and change all the only_if conditions. But if I could replace the os.redhat? conditional with selinux.installed? then I only need to change the test that ensures it gets installed on the correct systems. Once it's installed, the rest of the tests would automatically kick in and test it.

Not sure how much work is involved to make that happen, so I'm unable to determine if the benefit to it is worth the candle. Call it an edge case and if it's easily enough done, toss it in?

**Comment by adamleff** [comment]
@Paladin once it's a resource, and it has the right methods/matchers created, you can easily use it in `only_if` statements.

For example, if you could write a test like this:

```ruby
describe selinux do
  it { should be_installed }
end
```

... you'd be able to do this in a control:

```ruby
only_if { selinux.installed? }
```

It just works like magic :)

**Comment by Paladin** [comment]
Oooooh. (Don't do things like that to me! It's really hard to get drool out of this keyboard.)

**Comment by xianlin** [comment]
looks like this "selinux" resouce feature is still on the way and not implemented yet.

My question is that if I just want to know if my selinux configuration is "disabled" or not, can I just use current "File" resource and read the "content" of it to verify?

Or can I just use "inspec" to read the linux command output and verify it like "command $(getenforce) result should be "disabled""?

Or do you have any other work arounds? Thanks

**Comment by Paladin** [comment]
The command resource is exactly how I've been doing it, along with the package resource to test if it's installed.

**Comment by kmf** [comment]
Would love this!

**Comment by luckylittle** [comment]
+1

**Comment by predominant** [comment]
Yearly check-in 😃

**Comment by Schwad** [comment]
@predominant 

Hello! 👋 

I started working the the InSpec folks a few months ago, so this is my first time checking in here! Looks like the needs here are really well documented and there's some interest. 

I have added another label. We use this label whenever we have an issue that is clear, well defined and valuable (shovel-ready?) but unfortunately isn't directly in front of the team to be immediately consumed. I should come up with a better name for that label, but it's one we're actively sharing with #inspec channel on the community slack. 

This is more involved than a first issue, but there are some folks out there with great experience composing their own resources that I want to make sure see this issue. 

https://chefcommunity.slack.com/archives/C1XB6U6MN
